### PR TITLE
account encrypt/decrypt

### LINKED ISSUE #4474: Encrypt decrypt private key
- `encrypt` : Encrypts a private key to the web3 keystore v3 standard.

- `decrypt` :  Decrypts a keystore v3 JSON, and creates the account.

### LINKED ISSUE #4476: Wallet encrypt and decrypt
`encrypt`: Encrypts all wallet accounts to an array of encrypted keystore v3 objects. Param is password.

`decrypt`: Decrypts keystore v3 objects. params are keystoreArray and password

### PR DESCRIPTION
**Pull request by luu-alex**
#4476
Future TODO: add validation for keystore

Wallet/account Functions for: 
 - encrypt: Encrypts all wallet accounts to an array of encrypted keystore v3 objects. Param is password.
 - decrypt: Decrypts keystore v3 objects. params are keystoreArray and password

Changes
 - replaced dependancies `uuid` `scrypt-js` `crypto-browserify` with `ethereum-cryptography` library
 
Resources
https://github.com/ethereum/js-ethereum-cryptography#kdfs-pbkdf2-scrypt
https://github.com/ethereum/wiki/wiki/Web3-Secret-Storage-Definition

### CODE REVIEW

**Comment by luu-alex** [comment]
I'm running into a circular flow, where linter is complaining 
`web3.js/packages/web3-eth-accounts/src/account.ts
  82:73  error  'encrypt' was used before it was defined  no-use-before-define`
This is because, [privateKeyToAccount](https://github.com/ChainSafe/web3.js/pull/4622/files#diff-bcc6cd6da9c08aabb59251b6a4569eec1313010d9a1d187a03bbbe4118dcc648R56) returns includes the `encrypt` function.
And [encrypt](https://github.com/ChainSafe/web3.js/pull/4622/files#diff-bcc6cd6da9c08aabb59251b6a4569eec1313010d9a1d187a03bbbe4118dcc648R140) also uses `privateKeyToAccount`

File: package.json [hunk 1]
```diff
@@ -55,5 +55,6 @@
 		"ts-node": "^10.4.0",
 		"typescript": "^4.5.2"
 	},
-	"packageManager": "yarn@1.22.15"
+	"packageManager": "yarn@1.22.15",
+	"dependencies": {}
```
Comment (nazarhussain) [inline package.json]: Please avoid any unnecessary changes.

File: packages/web3-eth-accounts/src/types.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    salt: Buffer | string;
+}
+export type PBKDF2SHA256Params= {
+    c: number; // iterations
+    dklen: number;
+    prf: "hmac-sha256";
+    salt: Buffer | string;
+}
+
+export type V3Keystore = {
```
Comment (nazarhussain) [inline packages/web3-eth-accounts/src/types.ts]: Why its called `V3Keystore`? Can't we have more generic name?

File: packages/web3-eth-accounts/src/account.ts [hunk 3]
```diff
@@ -55,17 +80,146 @@ export const privateKeyToAccount = (
...
+    validateBytesInput(iv);
+
+    const kdf = options?.kdf ?? 'scrypt';
+
+    let derivedKey;
+    let kdfparams: ScryptParams | PBKDF2SHA256Params;
+
+    // derive key from key derivation function
+	if (kdf === 'pbkdf2') {
+        kdfparams = {
```
Comment (nazarhussain) [inline packages/web3-eth-accounts/src/account.ts]:
```suggestion
        kdfParams = {
```

File: packages/web3-eth-accounts/src/account.ts [hunk 4]
```diff
@@ -55,17 +80,146 @@ export const privateKeyToAccount = (
...
+    const seed = await createDecipheriv(Buffer.from(json.crypto.ciphertext, 'hex'), derivedKey.slice(0, 16), Buffer.from(json.crypto.cipherparams.iv, 'hex'));
+  
+    return privateKeyToAccount(Buffer.from(seed));
+};
+
+/**
+ * encrypt a private key given a password, returns a V3 JSON Keystore
+ * https://github.com/ethereum/wiki/wiki/Web3-Secret-Storage-Definition
+ */
+export const encrypt = async (privateKey: string, password: string | Buffer, options?: CipherOptions): Promise<V3Keystore> => {
```
Comment (nazarhussain) [inline packages/web3-eth-accounts/src/account.ts]: Should we set a minimum value for the `CipherOptions.c` a.k.a Iterations. `pbkdf2` is not secure less number of iterations. 
https://security.stackexchange.com/questions/3959/recommended-of-iterations-when-using-pbkdf2-sha256

**Review by nazarhussain — CHANGES_REQUESTED** [review]

**Comment by nazarhussain** [comment]
@luu-alex Use `eslint-disable-next-line` to avoid that no use before defined issue if can't be avoided.

File: packages/web3-eth-accounts/src/account.ts (hunk 3 above)
Comment (luu-alex) [reply packages/web3-eth-accounts/src/account.ts]: i stuck with all lowercase because `kdfparams` is how it is in 1.x, otherwise it would be considered a breaking change

File: packages/web3-eth-accounts/src/types.ts (hunk 2 above)
Comment (luu-alex) [reply packages/web3-eth-accounts/src/types.ts]: Changed to `KeyStore`

File: packages/web3-eth-accounts/src/account.ts (hunk 4 above)
Comment (luu-alex) [reply packages/web3-eth-accounts/src/account.ts]: That makes sense, I'll create validation to ensure `c` > 1000

File: packages/web3-common/src/errors.ts [hunk 5]
```diff
@@ -409,3 +410,17 @@ export class InvalidPrivateKeyError extends Web3Error {
...
+	public code = ERR_UNSUPPORTED_KDF;
+	public constructor(){
+		super(`Error: invalid key derivation function`)
+	}
+}
+
+export class KeyDerivationError extends Web3Error {
+	public code = ERR_UNSUPPORTED_KDF;
+	public constructor() {
+		super(`Error: Key derivation failed - possibly wrong password`)
```
Comment (Spacesai1or) [inline packages/web3-common/src/errors.ts]:
```suggestion
		super(`Invalid key derivation function`)
	}
}

export class KeyDerivationError extends Web3Error {
	public code = ERR_UNSUPPORTED_KDF;
	public constructor() {
		super(`Key derivation failed - possibly wrong password`)
```

No need to include `Error:` prefix, no other error message has it and JavaScript will already include it in native Error message

File: packages/web3-eth-accounts/src/account.ts [hunk 6]
```diff
@@ -1,35 +1,63 @@
...
+	KeyDerivationError,
+} from 'web3-common';
+import { Keystore, ScryptParams, PBKDF2SHA256Params, CipherOptions } from './types';
 
-// TODO Will be added later
-export const encrypt = (): boolean => true;
+const validateKeyStore = (keyStore: Keystore | string): boolean => !!keyStore;
 
-// TODO Will be added later
+// Will be added later
```
Comment (Spacesai1or) [inline packages/web3-eth-accounts/src/account.ts]:
```suggestion
// TODO Will be added later
```

File: packages/web3-eth-accounts/src/account.ts [hunk 7]
```diff
@@ -1,35 +1,63 @@
...
 export const sign = (): boolean => true;
 
-// TODO Will be added later
+// Will be added later
```
Comment (Spacesai1or) [inline packages/web3-eth-accounts/src/account.ts]:
```suggestion
// TODO Will be added later
```

**Comment by Spacesai1or** [comment]
[This](https://github.com/ChainSafe/web3.js/blob/265c0fd4ef9941c07e6e00c64e80a516b4b552dc/packages/web3-eth-accounts/src/account.ts#L73) can be updated now with [this](https://github.com/ChainSafe/web3.js/blob/4.x/packages/web3-utils/src/validation.ts#L447)

File: packages/web3-eth-accounts/src/account.ts [hunk 8]
```diff
@@ -53,10 +81,129 @@ export const privateKeyToAccount = (
 	const publicKeyString = `0x${publicKey.slice(2)}`;
```
Comment (Spacesai1or) [inline packages/web3-eth-accounts/src/account.ts:84]: Can we add a comment about why `.slice(2)`?

File: packages/web3-eth-accounts/src/account.ts [hunk 9]
```diff
@@ -53,10 +81,129 @@ export const privateKeyToAccount = (
...
+export const encrypt = async (
+	privateKey: HexString,
+	password: string | Buffer,
+	options?: CipherOptions,
+): Promise<Keystore> => {
+	if (!(isValidString(privateKey) || isBuffer(privateKey))) {
+		throw new InvalidPrivateKeyError(privateKey);
+	}
+
+	// if given salt or iv is a string, convert it to a Uint8Array
```
Comment (spacesailor24) [inline packages/web3-eth-accounts/src/account.ts:100]: is `iv` a typo?

File: packages/web3-eth-accounts/src/account.ts [hunk 10]
```diff
@@ -53,10 +81,129 @@ export const privateKeyToAccount = (
...
+	let salt;
+	if (options?.salt) {
+		salt = typeof options.salt === 'string' ? Buffer.from(options.salt, 'hex') : options.salt;
+	} else {
+		salt = randomBytes(32);
+	}
+	validateBytesInput(salt);
+
+	let iv;
```
Comment (Spacesai1or) [inline packages/web3-eth-accounts/src/account.ts]: Could `iv` get a more descriptive name, or is that what it's expected to be called?

File: packages/web3-eth-accounts/src/account.ts [hunk 11]
```diff
@@ -53,10 +81,129 @@ export const privateKeyToAccount = (
...
+		};
+		derivedKey = pbkdf2Sync(
+			Buffer.from(password),
+			Buffer.from(salt),
+			kdfparams.c,
+			kdfparams.dklen,
+			'sha256',
+		);
+	} else if (kdf === 'scrypt') {
+		// FIXME: support progress reporting callback
```
Comment (Spacesai1or) [inline packages/web3-eth-accounts/src/account.ts]:
```suggestion
		// TODO: support progress reporting callback
```

Are we actually wanted to support a callback here? We are trying to get rid of all callback if possible

File: packages/web3-eth-accounts/src/account.ts [hunk 12]
```diff
@@ -53,10 +81,129 @@ export const privateKeyToAccount = (
...
+/**
+ * encrypt a private key given a password, returns a V3 JSON Keystore
+ * https://github.com/ethereum/wiki/wiki/Web3-Secret-Storage-Definition
+ */
+export const encrypt = async (
+	privateKey: HexString,
+	password: string | Buffer,
+	options?: CipherOptions,
+): Promise<Keystore> => {
+	if (!(isValidString(privateKey) || isBuffer(privateKey))) {
```
Comment (Spacesai1or) [inline packages/web3-eth-accounts/src/account.ts:99]:
```suggestion
	if (!isValidString(privateKey) || isBuffer(privateKey)) {
```
It looks like this method is expected `privateKey` to only be a `HexString`. We need to either change the if statement as above, or change the type to allow buffers and handle converting them via:

```typescript
const stringPrivateKey = Buffer.isBuffer(privateKey)
		? Buffer.from(privateKey).toString('hex')
		: privateKey;
```

I would suggest following the same logic as `privateKeyToAddress`:

```typescript
const privateKeyToAddress = (privateKey: string | Buffer): string => {
	if (!(isValidString(privateKey) || isBuffer(privateKey))) {
		throw new InvalidPrivateKeyError(privateKey);
	}

	const stringPrivateKey = Buffer.isBuffer(privateKey)
		? Buffer.from(privateKey).toString('hex')
		: privateKey;

	const stringPrivateKeyNoPrefix = stringPrivateKey.startsWith('0x')
		? stringPrivateKey.slice(2)
		: stringPrivateKey;

	// TODO Replace with isHexString32Bytes function in web3-eth PR:
	// Must be 64 hex characters
	if (stringPrivateKeyNoPrefix.length !== 64) {
		throw new PrivateKeyLengthError(stringPrivateKeyNoPrefix);
	}
// ...
```

File: packages/web3-eth-accounts/src/account.ts [hunk 13]
```diff
@@ -53,10 +81,129 @@ export const privateKeyToAccount = (
...
+	// derive key from key derivation function
+	if (kdf === 'pbkdf2') {
+		kdfparams = {
+			dklen: options?.dklen ?? 32,
+			salt: salt.toString('hex'),
+			c: options?.c ?? 262144,
+			prf: 'hmac-sha256',
+		};
+		derivedKey = pbkdf2Sync(
+			Buffer.from(password),
```
Comment (Spacesai1or) [inline packages/web3-eth-accounts/src/account.ts]: `password` can be `string | Buffer` needs coercion and validation like `options.salt`:

```typescript
// if given salt or iv is a string, convert it to a Uint8Array
	let salt;
	if (options?.salt) {
		salt = typeof options.salt === 'string' ? Buffer.from(options.salt, 'hex') : options.salt;
	} else {
		salt = randomBytes(32);
	}
	validateBytesInput(salt);
```

File: packages/web3-eth-accounts/src/account.ts [hunk 14]
```diff
@@ -77,3 +224,78 @@ export const create = (): {
...
+	encrypt: (a: string, b: string) => Promise<Keystore>;
+}> => {
+	const json =
+		typeof keystore === 'object'
+			? keystore
+			: (JSON.parse(nonStrict ? keystore.toLowerCase() : keystore) as Keystore);
+
+	validateKeyStore(json);
+
+	if (json.version !== 3) throw new Error('Not a valid V3 wallet');
```
Comment (Spacesai1or) [inline packages/web3-eth-accounts/src/account.ts]: Replace with `Web3Error`

File: packages/web3-eth-accounts/src/account.ts [hunk 15]
```diff
@@ -77,3 +224,78 @@ export const create = (): {
...
+
+	let derivedKey;
+	if (json.crypto.kdf === 'scrypt') {
+		const kdfparams = json.crypto.kdfparams as ScryptParams;
+
+		// TODO support progress reporting callback
```
Comment (Spacesai1or) [inline packages/web3-eth-accounts/src/account.ts]: Again, are we sure if we need to support callback

File: packages/web3-eth-accounts/src/account.ts [hunk 16]
```diff
@@ -77,3 +224,78 @@ export const create = (): {
...
 	};
 };
+
+/**
+ *  Decrypts a v3 keystore JSON, and creates the account.
+ *
+ * */
+export const decrypt = async (
+	keystore: Keystore | string,
+	password: string,
```
Comment (spacesailor24) [inline packages/web3-eth-accounts/src/account.ts:234]: We should be uniform in our support of string and buffers, so we should probably add the ability to pass a buffer here

Comment (Spacesai1or) [inline packages/web3-eth-accounts/src/account.ts]: We should be uniform in our support for buffers, we should handle a buffer being passed here as we do for `encrypt`:

```typescript
export const encrypt = async (
	privateKey: HexString,
	password: string | Buffer,
// ...
```

File: packages/web3-eth-accounts/src/account.ts (hunk 8 above)
Comment (luu-alex) [reply packages/web3-eth-accounts/src/account.ts:84]: Unfortunately there are no comments on either [ethers](https://github.com/MaiaVictor/eth-lib/blob/da0971f5b09964d9c8449975fa87933f0c9fef35/lib/account.js#L25) or the library we were using before ([eth-lib](https://github.com/MaiaVictor/eth-lib/blob/da0971f5b09964d9c8449975fa87933f0c9fef35/lib/account.js#L25)). I can add these links as a comment to this line though

File: packages/web3-eth-accounts/src/errors.ts [hunk 17]
```diff
@@ -0,0 +1,27 @@
+/* eslint-disable max-classes-per-file */
+
+export abstract class Web3Error extends Error {
```
Comment (Spacesai1or) [inline packages/web3-eth-accounts/src/errors.ts]: `Web3Error` should be imported from [web3-common](https://github.com/ChainSafe/web3.js/blob/4.x/packages/web3-common/src/errors.ts#L45) and not re-declared

File: packages/web3-eth-accounts/test/fixtures/account.ts [hunk 18]
```diff
@@ -29,3 +31,186 @@ export const invalidPrivateKeytoAccountData: [any, string][] = [
 	[undefined, 'Invalid value given "undefined". Error: not a valid string or buffer.'],
 	[null, 'Invalid value given "null". Error: not a valid string or buffer.'],
 ];
+
+export const validEncryptData: [[any, string, CipherOptions], any][] = [
```
Comment (Spacesai1or) [inline packages/web3-eth-accounts/test/fixtures/account.ts]: `validEncryptData: [[any, string, CipherOptions], any][]`

should be:

`validEncryptData: [[HexString, string | Buffer, CipherOptions], Keystore][]`

File: packages/web3-eth-accounts/test/fixtures/account.ts [hunk 19]
```diff
@@ -29,3 +31,186 @@ export const invalidPrivateKeytoAccountData: [any, string][] = [
...
+					c: 262144,
+					prf: 'hmac-sha256',
+				},
+				mac: '46eb4884e82dc43b5aa415faba53cc653b7038e9d61cc32fd643cf8c396189b7',
+			},
+		},
+	],
+];
+
+export const invalidEncryptData: [[any, string, any], Error][] = [
```
Comment (Spacesai1or) [inline packages/web3-eth-accounts/test/fixtures/account.ts]:
```suggestion
export const invalidEncryptData: [[any, string, any], PrivateKeyLengthError | InvalidKdfError][] = [
```

File: packages/web3-eth-accounts/test/fixtures/account.ts [hunk 20]
```diff
@@ -29,3 +31,186 @@ export const invalidPrivateKeytoAccountData: [any, string][] = [
...
+				iv: 'bfb43120ae00e9de110f8325143a2709',
+				salt: '210d0ec956787d865358ac45716e6dd42e68d48e346d795746509523aeb477dd',
+				kdf: 'hkdf',
+			},
+		],
+		new InvalidKdfError(),
+	],
+];
+
+export const validDecryptData: [[any, string, CipherOptions]][] = [
```
Comment (Spacesai1or) [inline packages/web3-eth-accounts/test/fixtures/account.ts]: `validDecryptData: [[any, string, CipherOptions]][]`

should be

`validDecryptData: [[Keystore | string, string, CipherOptions]][]`

Also should have test cases for [nonStrict](https://github.com/ChainSafe/web3.js/blob/265c0fd4ef9941c07e6e00c64e80a516b4b552dc/packages/web3-eth-accounts/src/account.ts#L235)

File: packages/web3-eth-accounts/test/fixtures/account.ts [hunk 21]
```diff
@@ -29,3 +31,186 @@ export const invalidPrivateKeytoAccountData: [any, string][] = [
...
+				salt: Buffer.from(
+					'210d0ec956787d865358ac45716e6dd42e68d48e346d795746509523aeb477dd',
+					'hex',
+				),
+			},
+		],
+	],
+];
+
+export const invalidDecryptData: [[any, string], any][] = [
```
Comment (Spacesai1or) [inline packages/web3-eth-accounts/test/fixtures/account.ts]:
```suggestion
export const invalidDecryptData: [[any, string], InvalidKdfError | KeyDerivationError][] = [
```

File: packages/web3-eth-accounts/src/account.ts (hunk 11 above)
Comment (luu-alex) [reply packages/web3-eth-accounts/src/account.ts]: This was a line from 1.x, this seems outdated so I am in favor of removing this

**Comment by Spacesai1or** [comment]
I think the invalid test data should also test for missing parameters e.g. the [invalidEncryptData](https://github.com/ChainSafe/web3.js/blob/265c0fd4ef9941c07e6e00c64e80a516b4b552dc/packages/web3-eth-accounts/test/fixtures/account.ts#L127) should test for failures if `keystore` and `password` are not provided at all

File: packages/web3-utils/src/validation.ts [hunk 22]
```diff
@@ -37,6 +37,21 @@ export const validateHexStringInput = (data: HexString) => {
 	}
 };
 
+/**
+ * 
+ */
+ export function assertBytes(bytes: Uint8Array, ...len: number[]) {
```
Comment (Spacesai1or) [inline packages/web3-utils/src/validation.ts]: Needs tests

**Review by Spacesai1or — CHANGES_REQUESTED** [review]
Great job tackling the cryptography related stuff!

File: packages/web3-utils/src/validation.ts (hunk 22 above)
Comment (luu-alex) [reply packages/web3-utils/src/validation.ts]: removed, didn't end up using the method

File: packages/web3-eth-accounts/src/account.ts (hunk 8 above)
Comment (Spacesai1or) [reply packages/web3-eth-accounts/src/account.ts:84]: Yea, can't find anything about either

**Review by nazarhussain — APPROVED** [review]

File: packages/web3-eth-accounts/test/fixtures/account.ts [hunk 23]
```diff
@@ -23,9 +33,229 @@ export const validPrivateKeytoAccountData: [string, any][] = [
 	],
 ];
 
-export const invalidPrivateKeytoAccountData: [any, string][] = [
-	['', 'Invalid value given "". Error: Private key must be 32 bytes.'],
-	[Buffer.from([]), 'Invalid value given "". Error: Private key must be 32 bytes.'],
-	[undefined, 'Invalid value given "undefined". Error: not a valid string or buffer.'],
-	[null, 'Invalid value given "null". Error: not a valid string or buffer.'],
+export const invalidPrivateKeytoAccountData: [any, any][] = [
```
Comment (Spacesai1or) [inline packages/web3-eth-accounts/test/fixtures/account.ts]:
```suggestion
export const invalidPrivateKeytoAccountData: [any, PrivateKeyLengthError | InvalidPrivateKeyError][] = [
```

File: packages/web3-eth-accounts/test/fixtures/account.ts [hunk 24]
```diff
@@ -23,9 +33,229 @@ export const validPrivateKeytoAccountData: [string, any][] = [
...
+	['', new PrivateKeyLengthError()],
+	[Buffer.from([]), new PrivateKeyLengthError()],
+	[undefined, new InvalidPrivateKeyError()],
+	[null, new InvalidPrivateKeyError()],
+];
+
+export const validEncryptData: [[any, string | Buffer, CipherOptions], any][] = [
```
Comment (Spacesai1or) [inline packages/web3-eth-accounts/test/fixtures/account.ts]:
```suggestion
export const validEncryptData: [[any, string | Buffer, CipherOptions], KeyStore][] = [
```

File: packages/web3-eth-accounts/test/fixtures/account.ts [hunk 25]
```diff
@@ -23,9 +33,229 @@ export const validPrivateKeytoAccountData: [string, any][] = [
...
+					c: 262144,
+					prf: 'hmac-sha256',
+				},
+				mac: '46eb4884e82dc43b5aa415faba53cc653b7038e9d61cc32fd643cf8c396189b7',
+			},
+		},
+	],
+];
+
+export const invalidEncryptData: [[any, any, any], PrivateKeyLengthError | InvalidKdfError][] = [
```
Comment (Spacesai1or) [inline packages/web3-eth-accounts/test/fixtures/account.ts]:
```suggestion
export const invalidEncryptData: [[any, any, any], PrivateKeyLengthError | InvalidKdfError | InvalidPrivateKeyError | InvalidPasswordError | IVLengthError | PBKDF2IterationsError][] = [
```

File: packages/web3-utils/src/validation.ts [hunk 26]
```diff
@@ -122,12 +122,12 @@ export const validateNumbersInput = (
 /**
  * checks input if typeof data is valid string input
  */
-export const isValidString = (data: any) => typeof data === 'string';
+export const isValidString = (data: Bytes) => typeof data === 'string';
```
Comment (Spacesai1or) [inline packages/web3-utils/src/validation.ts]:
```suggestion
export const isValidString = (data: unknown) => typeof data === 'string';
```
For validation methods, we don't know the type of data being passed

File: packages/web3-utils/src/validation.ts [hunk 27]
```diff
@@ -122,12 +122,12 @@ export const validateNumbersInput = (
...
 
 /**
  * checks input if typeof data is valid buffer input
  */
-export const isBuffer = (data: any) => Buffer.isBuffer(data);
+export const isBuffer = (data: Bytes) => Buffer.isBuffer(data);
```
Comment (Spacesai1or) [inline packages/web3-utils/src/validation.ts]:
```suggestion
export const isBuffer = (data: unknown) => Buffer.isBuffer(data);
```

File: packages/web3-utils/src/validation.ts [hunk 28]
```diff
@@ -444,6 +444,13 @@ export const validateHexString8Bytes = (value: HexString8Bytes, prefixed = true)
 	if (!isHexString8Bytes(value, prefixed)) throw new InvalidHexStringError(value, 8);
 };
 
+export const isHexString16Bytes = (value: HexString32Bytes, prefixed = true) =>
```
Comment (Spacesai1or) [inline packages/web3-utils/src/validation.ts]:
```suggestion
export const isHexString16Bytes = (value: unknown, prefixed = true) =>
```

File: packages/web3-utils/src/validation.ts [hunk 29]
```diff
@@ -444,6 +444,13 @@ export const validateHexString8Bytes = (value: HexString8Bytes, prefixed = true)
...
+	prefixed ? isHexStrict(value) && value.length === 34 : isHex(value) && value.length === 32;
+
+export const validateHexString16Bytes = (value: HexString32Bytes, prefixed = true) => {
```
Comment (Spacesai1or) [inline packages/web3-utils/src/validation.ts]:
```suggestion
export const validateHexString16Bytes = (value: unknown, prefixed = true) => {
```

**Review by Spacesai1or — APPROVED** [review]
